### ISSUE TITLE
Option to disable suggestions initially

### ISSUE THREAD

**Issue by kevintraver** [toranb/ember-cli-auto-complete#16]
It might be nice to have some sort of way to disable the results initially until a search has been entered.

Thanks!

**Comment by kevintraver** [comment]
Ill work on a pull request for this. It might also be nice to have a minimum search length property.

**Comment by toranb** [comment]
Agreed on both features :)

Would love to turn that off (optionally) and if adding a min search length achieves this - even better!

Thanks for offering to do the PR yourself! The test suite is comprehensive so it should provide a good working example of how to prove the feature works as expected (without breaking any of the existing behavior)

Let me know if you have questions or need any help!

**Comment by stravid** [comment]
I'm looking into implementing this feature myself and already read the codebase. Before I start I wanted to ask a few questions so we are on the same page @toranb:
- Currently one has to implement a computed property to determine the suggested options. This solution is tightly coupled to the internals of the component, namely it depends on the `inputVal` and `options` attributes. I suggest moving this into a "dumb" function so it is decoupled from the components internals.
  
  Current:
  
  ``` javascript
  suggestions: function() {
    var inputVal = this.get("inputVal") || "";
    return this.get("options").filter(function(item) {
        return item.get("code").toLowerCase().indexOf(inputVal.toLowerCase()) > -1;
    });
  }.property("inputVal", "options.@each"),
  ```
  
  Proposed change:
  
  ``` javascript
  determineSuggestions: function(options, inputValue) {
    return options.filter(function(item) {
        return item.get("code").toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
    });
  },
  ```
  
  This would also allow the component to control when this "expensive" task is run. When the input value is not above the minimum length treshold it's not clever to determine the suggestions. With the current solution it is not so easy to prevent the computed property from doing that.
- Is there a specific reason why the suggestions are shown / hidden via an inline style attribute and not a Handlebars `if` helper?
- Is there a convention for the coding style? Currently it looks a little random, I don't mind making in coherent.

**Comment by toranb** [comment]
Sorry for the massive delay here - on the proposed change -> looks much cleaner than what I had for sure. Assuming the tests pass I'd call that a win/win

About the inline style attrs - I'd love a hbs helper if you don't mind rolling one /updating this to use it. Keep in mind this is built for ember 1.12+ so use the latest hbs helper syntax if you would please :)

If you want to cleanup the code - the only thing I usually keep w/ is 4 spaces (as I'm also a python dev and it keeps me in sync) but ... that said anything else should be cool.

You sound like you really want to improve this /would you like a commit bit also?

thanks for the awesome suggestions/thoughtful issue before you fire off a PR !!!

**Comment by stravid** [comment]
No worries, I know how time demanding open source is and that there are other, more important things in life :)

I already finished the refactoring from implicit computed properties to explicit functions with all tests passing. While doing so I found a few things that are not entirely clear to me, maybe you can shed some light on it:
1. What is the purpose of `optionsToMatch`?
2. What is the difference between `selectedValue` and `inputVal`? Why are there both if one could be enough?

As you can see, I struggle a little to understand the use cases and why certain things are the way they are. Regarding the commit bit, let me earn it first ;)

**Comment by toranb** [comment]
1) I think those are literally the "options" that show up as you typeahead (think select -these would literally be options)
2) I think selectedValue might be the only part that matters (it's the basic value tracking what the user inputs from what I can tell). If you can eliminate one and still have it work I'd call that a win

**Comment by stravid** [comment]
I hope it's okay for you if I ask so many questions, I'm trying to understand the component before I try to improve things. Regarding the `optionsToMatch`: I stepped through every imaginable scenario with the debugger and I think it is used for a single feature:

If the input loses focus and the entered text is an exact match to an option the input is not cleared.

The user supplied functions `suggestions` and `optionsToMatch` are almost the same. I would like to drop the `optionsToMatch` function and thereby reduce the API surface. To do this we will use the `suggestions` to determine if the input should be cleared. Now it get's tricky because I would like to remove a quirk of the current code.

Given the following options here is what happens when the input loses focus.
- ABC
- ABD
- BCD

| Input | # of visible Suggestions | Current Behaviour | New Behaviour 1 | New Behaviour 2 |
| --- | --- | --- | --- | --- |
| A | 2 | Clear input | Clear input | Clear input |
| ABC | 1 | Keep input | Keep input | Keep input |
| abc | 1 | Keep input | Keep input | Keep input |
| Abc | 1 | Clear input | Keep input | Keep input |
| CD | 1 | Clear input | Clear input | Keep input |

The quirk is in the next to last row. `Abc` does not match any of the options because `toLowerCase` is only called on one side of the comparison therefore the input is cleared. I think this is weird.

I propose one of two changes

a.) Lowercase both sides and keep the input if there is an exact match. This removes the mentioned quirk. It does change behaviour but since there is no test covering this I assume it is not an official feature.

b.) This changes more and breaks the current tests but simplifies the logic and explanation of what is happening: When the input loses focus and there is only one suggestion left keep the text, otherwise clear it.

What do you think?

**Comment by toranb** [comment]
Wow this was a serious look at the addon/behavior and I must say I'm impressed :)

About the table -just to clarify your commands

1) clear input meaning you wipe the input box and blur away
2) keep input meaning you hit enter or ? does it instead mean you blur away (tab key) ?

About the bug/issue you mention "Abc" + blur I can see your point of view and agree it's an improvement. From the testing side it would appear you could copy/paste this test + update it to show the bug in question

https://github.com/toranb/ember-cli-auto-complete/blob/master/tests/acceptance/end-to-end-test.js#L50-L64

Thanks again for looking at this from a fresh perspective!

**Comment by stravid** [comment]
Regarding the table: The blur happens before the table comes into play. The table visualises what happens when a blur is triggered. When then users hits tab / enter / clicks outside the, component tries to decide what should happen with the inputs text.

Clear input = The component resets the inputs text to an empty string
Keep input = The component doesn't touch the input text

You are absolutely right, a new test like the one you highlighted would show the quirk.

**Comment by toranb** [comment]
@stravid should I still reply to your comments above before we do a 1.0 release so we can get this inline with your expectations?

**Comment by stravid** [comment]
@toranb If you have the time, sure!

**Comment by toranb** [comment]
So I finally took the 2 minutes to read your 2 suggestions :) and I think I like option a right now because it _should_ toLower both sides and honestly this is a bug IMHO.

The 2nd option is interesting though because in my "programming" experience it acts more like a fuzzy finder (ie- how I use vim day-to-day now). Maybe you could explain this a bit more to help me get on board :) after all we are about to bump this to v1.0.0 so anything goes right

I'm mostly confused (a bit) by the example of CD - just to clarify this is "matching" the BCD option found correct? The only tests that break would be ones that "strictly" match (not fuzzy like we would migrate toward).

I _think_ I'm 99% w/ you on this because most select replacements already have this behavior. Select 2 (example below) is just one proof 

https://istefo.github.io/ember-select-2/#/examples

**Comment by stravid** [comment]
My pull request already implemented option a. It also added a regression test to make clear that this behaviour was a bug. (I'm with you on this one.)

I can understand the confusion, it took me a while to wrap my head around it. Here is my current explanation what happens inside the component:

The logic for determining the suggestions acts more or less like a fuzzy finder, depending on the consumers implementation.

The other logical part of the component, "what to do when a blur is triggered", though has it's very own logic to determine if the input should be cleared. Compared to the fuzzy find logic of the suggestions it operates with a equality logic.

This difference in logic causes the, in my opinion weird, behaviour we see:

`CD` fuzzy matches `BCD` therefore it is listed as suggestion. But when the input is blurred it gets cleared because it is not an exact match. That's the result of having this two different types of logic in the component.

So my proposition is to treat the consumer implemented fuzzy find logic as the one source of truth. We clear the input if there are more than one suggestions. Otherwise we keep it because the fuzzy find logic already determined that there is only one suggestion left matching, why not keep it in this case.

And you are totally correct, the only tests that would break are the ones that test the "strictly" matching logic.

I hope now it's more clear to you.

**Comment by toranb** [comment]
I'm coming around to the idea as part of the v1.0.0 release (as some might see this as a breaking change). I've played around with a few select options this past week and every single one does exactly what you described above.

Did you already have a PR /something in flight that would update this to match more like a real fuzzy matcher would or is this a feature request of sorts?

**Comment by stravid** [comment]
It's not yet implemented but my last pull request made it easy to switch to this logic. I will implement it and open a new pull request if that works for you.

**Comment by toranb** [comment]
@stravid would love that :) but no rush on my part. I plan to wait on the 1.0 release until we get your latest PR and I get a chance to have @joe-l-bright verify it's solid in his pre-1.0 apps

**Comment by toranb** [comment]
I'm closing out this issue - let me know when you get time this month/or next to crank out the PR :)
